Settings edit button

shubham-padia opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Let's make the tooltip "Change language".

alya commented:
I'll let @sahil839 review before I test, but how is the size of the language button determined? Can you post screenshots with a really long and a really short language name, to confirm that everything looks good?

alya commented:
Let's try positioning the email and password buttons as described in the issue:

> Under Account & privacy, put the email in a button below its heading...
> For the password, just have a "Change your password" button just under the email (no field label, no ***s, no tooltip).

sahil839 commented:
I did not test this, but the code changes look good.

alya commented:
What's the answer to this quetion above?

> how is the size of the language button determined?

alya commented:
After I click the language button, and then click out of the language picker modal, the tooltip gets stuck:

![Screenshot 2024-07-03 at 11 43 58](https://github.com/zulip/zulip/assets/2090066/89bcce4a-a921-49d7-b86c-2b9bdbf81097)

alya commented:
The email also has this issue:

![Screenshot 2024-07-03 at 11 45 37](https://github.com/zulip/zulip/assets/2090066/40142085-de94-40ab-adf6-fbef94819beb)

shubham-padia commented:
I've fixed the tooltip getting stuck issue.

> but how is the size of the language button determined? Can you post screenshots with a really long and a really short language name, to confirm that everything looks good?

I've added both screenshots to the PR description. Minimum width of the button is 130px, this is the same button as `Manage your API key button`, which adds to the extra space that we see. For the other buttons in settings, since their string is usually long enough, we don't notice the extra space. 

This is how the button will look like if we keep the `.small` class we were using for this button already.

| English language | Short Language | Long language |
| -- | -- | -- |
| <img width="344" alt="Screenshot 2024-07-04 at 4 12 12 PM" src="https://github.com/zulip/zulip/assets/13910561/d51bbf54-c2ad-471c-8661-a1f452e671ee"> |<img width="344" alt="Screenshot 2024-07-04 at 4 12 04 PM" src="https://github.com/zulip/zulip/assets/13910561/ef204e73-c96e-4f60-9bc3-4cf4fc16df8e"> | <img width="344" alt="Screenshot 2024-07-04 at 4 11 55 PM" src="https://github.com/zulip/zulip/assets/13910561/4fed2932-327c-4d8d-9569-9fb5ceabd4a5">

This is how it will look like if reduce the min width to 90 px
| English language | Short Language | Long language |
| -- | -- | -- |
| <img width="344" alt="Screenshot 2024-07-04 at 4 16 26 PM" src="https://github.com/zulip/zulip/assets/13910561/3747ffc8-07c9-4dfe-9ac7-45d0abdf7f04"> | <img width="344" alt="Screenshot 2024-07-04 at 4 16 44 PM" src="https://github.com/zulip/zulip/assets/13910561/ef953654-1d07-4c43-9fb2-dad4065931d3"> |  <img width="344" alt="Screenshot 2024-07-04 at 4 16 35 PM" src="https://github.com/zulip/zulip/assets/13910561/64beb0f0-6c01-4caa-ab5e-2c3766ef74db">

sahil839 commented on web/src/click_handlers.js:
```diff
@@ -809,6 +809,9 @@ export function initialize() {
         e.stopPropagation();
+        // We need to trigger `blur` to make the tooltip go away after
+        // the click event.
+        $(e.currentTarget).trigger("blur");
```
At other places, this is fixed by setting the trigger for tooltip to `mouseenter`. 

Like we have `tippy-delayed-zulip-tooltip` class which does this but also adds a delay while showing the tooltip. So, this can be fixed adding code in `tippy.js` in a similar way.

shubham-padia replied:
Thanks for the suggestion, I've set the trigger to mouseenter.

shubham-padia commented:
This is ready for re-review btw!

alya commented:
> Give it a "Change our language" tooltip (with appropriate delay).

What delay is the tooltip using? It's definitely showing up too fast.

alya commented:
I think these buttons don't need a minimum width. Just a reasonable margin (whatever we normally use) should look fine.

alya commented:
If other buttons have a min width that has no effect, that might be a thing to clean up too, but I'll leave it to @sahil839 and @timabbott to comment.

shubham-padia commented:
> I think these buttons don't need a minimum width. Just a reasonable margin (whatever we normally use) should look fine; the extra white space around the text looks worse to me.

In that case, I'll use the `.small` class shown in the comment: https://github.com/zulip/zulip/pull/30697#issuecomment-2208503968

shubham-padia commented:
> What delay are the tooltips using? They are definitely showing up too fast.

I've fixed this up, the tooltip has the `LONG_HOVER_DELAY` now, which is 750 ms to appear and 20 ms to disappear.

For the visual changes requested, updated the screenshots in the PR.

sahil839 commented:
> If other buttons have a min width that has no effect, that might be a thing to clean up too, but I'll leave it to @sahil839 and @timabbott to comment.

I guess the min-width property was added to make the buttons not look too small for certain cases, so if we are happy with buttons just having width according to the text and some padding around the text, then it would be good to clean that up.

shubham-padia commented:
> I guess the min-width property was added to make the buttons not look too small for certain cases, so if we are happy with buttons just having width according to the text and some padding around the text, then it would be good to clean that up.

Yeah, but it would be good to do it in a seperate issue since we might have some buttons changing their appearance a little bit, when we remove the `min-width`. And it would be good to consider whether those changes look good to us or not in a separate context.

alya commented:
There's still something wrong here. At 16px, the "Manage your API" button scales up nicely, but these new buttons do not.

![Screenshot 2024-07-09 at 12 07 15](https://github.com/zulip/zulip/assets/2090066/85296199-1155-430f-9d1d-b955be267d22)

alya commented:
Please test carefully at both legacy scaling and 16px.

alya commented:
> Yeah, but it would be good to do it in a seperate issue since we might have some buttons changing their appearance a little bit, when we remove the min-width. And it would be good to consider whether those changes look good to us or not in a separate context.

Yeah, this is a fine task to leave for after the release.

alya commented:
I don't understand why we need to use `.small` here. Why not make them 14px/16px, like the API button, as suggested in the issue?

shubham-padia commented:
> Yeah, this is a fine task to leave for after the release. @shubham-padia can you add it to your todo list? You can file an issue if that will help keep track of it. The first step might be to post in #design with examples of what this would actually change visually.

Opened #30806

shubham-padia commented:
Removed `.small` and removed min-width for buttons that we are dealing with in this PR. We should remove the `min-width: 0` added here in #30806.

Updated the PR description with 14px and 16px screenshots

alya commented:
Actually, this begs the question: are we using .small on other buttons? Clearly those aren't going to scale up properly if we are.

alya commented:
I think these look fine both at 14px and 16px now. @sahil839 do you have any feedback here?

shubham-padia commented:
> Actually, this begs the question: are we using .small on other buttons? Clearly those aren't going to scale up properly if we are.

There's 58 occurrences currently for this. For some places like the compose box, we override the font-size for `.small` for the buttons used in compose box, which makes the problem go away. For other places like `Select all drafts` button in draft selector, the font size does not scale. I'll have a look at this later, see what the options to solve this are, how they look and start a CZO thread about this.

alya commented:
Got it, yeah, let's discuss on CZO.

alya commented:
@sahil839 any more feedback before passing this PR on for integration review? I think visually the latest version looks fine.

sahil839 commented:
Looks good.

timabbott commented:
Merged, thanks @shubham-padia!
